app/src/main/kotlin/com/ark/globe/coordinates/URLParser.kt

```diff
@@ -11,45 +8,72 @@ import java.net.URL
 class URLParser {
     companion object{
         private const val GOOGLE_MAPS = "maps.google.com"
+        private const val GOOGLE_MAPS_SHORT = "goo.gl"
         private const val OPENSTREETMAP = "www.openstreetmap.org"
         private const val OSM_AND_NET = "osmand.net"
         private const val URL_PROTOCOL_1 = "http"
         private const val URL_PROTOCOL_2 = "https"
+        private var coordinatesString = ""
+        private var latitude: String? = null
+        private var longitude: String? = null
+
+        private val ioDispatcher = Dispatchers.IO
 
         suspend fun extractCoordinates(url: String?): Coordinates?{
-            return withContext(Dispatchers.Default) {
-                var latitude: String? = null
-                var longitude: String? = null
+            var coordinates: Coordinates? = null
+            val validUrl = getValidURL(url)
 
-                if (url != null && url.contains(GOOGLE_MAPS)) {
-                    //google maps link processing
-                    val coordinatesString = url.substringAfter("?q=").substringBefore("&")
-                    latitude = coordinatesString.substringBefore(",")
-                    longitude = coordinatesString.substringAfter(",")
-                }
+            withContext(ioDispatcher) {
 
-                if (url != null && url.contains(OPENSTREETMAP)) {
+                parseGoogleMapsLinks(validUrl)
+
+                if (validUrl != null && validUrl.contains(OPENSTREETMAP)) {
                     //openstreetmap link processing
-                    val coordinatesString = url.substringAfter("#map").substringAfter("/")
+                    coordinatesString = validUrl.substringAfter("#map").substringAfter("/")
```

Comment:
The same, better to use some auxiliary lib for parsing query parameters, this will be easier to write, easier to read and more robust.

Reply:
Ok, I that is a better way indeed. We should look for a lib to do that. But do you think the query parameter keys are fixed?